### PR TITLE
[APInt] Fix APInt constructions where value does not fit bitwidth (NFCI)

### PR DESCRIPTION
**Pull request by nikic**
This fixes all the places that hit the new assertion added in https://github.com/llvm/llvm-project/pull/106524 in tests. That is, cases where the value passed to the APInt constructor is not an N-bit signed/unsigned integer, where N is the bit width and signedness is determined by the isSigned flag.

The fixes either set the correct value for isSigned, set the implicitTrunc flag, or perform more calculations inside APInt.

Note that the assertion is currently still disabled by default, so this patch is mostly NFC.

### CODE REVIEW

**Comment by nikic** [comment]
Before I spend time tracking down all assertion failures, I'd like to have some feedback on the general change.

File: llvm/include/llvm/ADT/APInt.h [hunk 1]
```diff
@@ -106,11 +106,26 @@ class [[nodiscard]] APInt {
...
   /// \param isSigned how to treat signedness of val
-  APInt(unsigned numBits, uint64_t val, bool isSigned = false)
+  /// \param implicitTrunc allow implicit truncation of non-zero/sign bits of
+  ///                      val beyond the range of numBits
+  APInt(unsigned numBits, uint64_t val, bool isSigned = false,
+        bool implicitTrunc = false)
       : BitWidth(numBits) {
+    if (!implicitTrunc) {
+      if (BitWidth == 0) {
+        assert(val == 0 && "Value must be zero for 0-bit APInt");
```
Comment (RKSimon) [inline llvm/include/llvm/ADT/APInt.h:116]: Is it worth putting the if-else conditions inside the assert directly?

**Review by RKSimon — COMMENTED** [review]
Is the plan to set implicitTrunc in so many places or is this part of tracking down all the regressions?

File: llvm/include/llvm/ADT/APInt.h (hunk 1 above)
Comment (nikic) [reply llvm/include/llvm/ADT/APInt.h:116]: As the conditions end up fairly complex, I figured it's better to have if/else. If assert() expands to nothing, they'll get optimized away.

**Comment by nikic** [comment]
> Is the plan to set implicitTrunc in so many places or is this part of tracking down all the regressions?

Apart from a couple marked TODO, most of them are here to stay. E.g. that bit of code in ConstantFolding has "sext or trunc" semantics, so doing something like "isSigned = true, implicitTrunc = true" there makes sense.

Some of them could be avoided, but it's not really clear that it's worthwhile/desirable. E.g. the one in FuzzMutate/OpDescriptor really doesn't really care about the actual value, so doing a truncation is as good as anything else.

**Comment by nikic** [comment]
As a side note, it seems like not having the implicit truncation also improves compile-time: http://llvm-compile-time-tracker.com/compare.php?from=fd191378dce6b20c100da716f94130af2593df37&to=df515f3e94d5f5d10ecddfb60181d8866817cc27&stat=instructions:u

**Comment by vladimirradosavljevic** [comment]
@nikic Thanks for working on this!
I discovered issues in downstream 256-bit target that are fixed by this PR, so are you planning to continue to work on this?
Thanks again for the fixes.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp [hunk 2]
```diff
@@ -2142,7 +2142,9 @@ ScheduleDAGSDNodes *SelectionDAGISel::CreateScheduler() {
 bool SelectionDAGISel::CheckAndMask(SDValue LHS, ConstantSDNode *RHS,
                                     int64_t DesiredMaskS) const {
   const APInt &ActualMask = RHS->getAPIntValue();
-  const APInt &DesiredMask = APInt(LHS.getValueSizeInBits(), DesiredMaskS);
+  // TODO: Avoid implicit trunc?
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:2145]: I think this issue exists in CheckOrMask too, but it isn't tested in tree.

**Comment by nikic** [comment]
Okay, looks like I tracked down all the assertion failures in tests. My plan would now be to land this in three phases, first to add the new assertions and implicitTrunc parameter, but with implicitTrunc=true by default. Then to land fixes (possibly split up further), and finally to flip the implicitTrunc default.

**Comment by nikic** [comment]
I think this is ready for review now. I've landed many parts of this PR separately already, and am happy to land this in parts (or split as needed).

File: llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp [hunk 3]
```diff
@@ -1806,7 +1806,7 @@ bool AMDGPUDAGToDAGISel::SelectGlobalSAddr(SDNode *N,
       // instructions to perform VALU adds with immediates or inline literals.
       unsigned NumLiterals =
           !TII->isInlineConstant(APInt(32, COffsetVal & 0xffffffff)) +
-          !TII->isInlineConstant(APInt(32, COffsetVal >> 32));
+          !TII->isInlineConstant(APInt(32, uint64_t(COffsetVal) >> 32));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:1809]: These should probably just use Lo_32/Hi_32

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 4]
```diff
@@ -4377,7 +4377,7 @@ AMDGPUInstructionSelector::selectGlobalSAddr(MachineOperand &Root) const {
         // instructions to perform VALU adds with immediates or inline literals.
         unsigned NumLiterals =
             !TII.isInlineConstant(APInt(32, ConstOffset & 0xffffffff)) +
-            !TII.isInlineConstant(APInt(32, ConstOffset >> 32));
+            !TII.isInlineConstant(APInt(32, uint64_t(ConstOffset) >> 32));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:4380]: These should probably just use Lo_32/Hi_32

**Comment by dtcxzyw** [comment]
https://github.com/llvm/llvm-project/blob/62f737f7409b5d2b33c746158c62f14e5bb78aed/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L6516
We should allow truncations here.

```
; bin/opt -passes='simplifycfg<switch-to-lookup>' reduced.ll -S
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define i1 @_ZN5image5color9ColorType9has_alpha17h81054606b63d0c15E(i8 %0) {
  switch i8 %0, label %3 [
    i8 9, label %2
    i8 1, label %2
    i8 7, label %2
    i8 3, label %2
    i8 5, label %2
  ]

2:                                                ; preds = %1, %1, %1, %1, %1
  br label %3

3:                                                ; preds = %2, %1
  %.0 = phi i1 [ true, %2 ], [ false, %1 ]
  ret i1 %.0
}
```
```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/include/llvm/ADT/APInt.h:121: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isUIntN(BitWidth, val) && "Value is not an N-bit unsigned value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=simplifycfg<switch-to-lookup> reduced.ll -S
1.      Running pass "function(simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;no-switch-range-to-icmp;switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>)" on module "reduced.ll"
2.      Running pass "simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;no-switch-range-to-icmp;switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>" on function "_ZN5image5color9ColorType9has_alpha17h81054606b63d0c15E"
 #0 0x0000717e69813f12 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x213f12)
 #1 0x0000717e69810ddf llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x210ddf)
 #2 0x0000717e69810f25 SignalHandler(int) Signals.cpp:0:0
 #3 0x0000717e69242520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x0000717e692969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x0000717e692969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x0000717e692969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x0000717e69242476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x0000717e692287f3 abort ./stdlib/abort.c:81:7
 #9 0x0000717e6922871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x0000717e69239e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x0000717e62d2afef llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMTransformUtils.so.20.0git+0x12afef)
#12 0x0000717e62e97856 (anonymous namespace)::SwitchLookupTable::SwitchLookupTable(llvm::Module&, unsigned long, llvm::ConstantInt*, llvm::SmallVectorImpl<std::pair<llvm::ConstantInt*, llvm::Constant*>> const&, llvm::Constant*, llvm::DataLayout const&, llvm::StringRef const&) SimplifyCFG.cpp:0:0
#13 0x0000717e62eb4a96 switchToLookupTable(llvm::SwitchInst*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&, llvm::DomTreeUpdater*, llvm::DataLayout const&, llvm::TargetTransformInfo const&) SimplifyCFG.cpp:0:0
#14 0x0000717e62ebfedd (anonymous namespace)::SimplifyCFGOpt::simplifySwitch(llvm::SwitchInst*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) SimplifyCFG.cpp:0:0
#15 0x0000717e62ec557b llvm::simplifyCFG(llvm::BasicBlock*, llvm::TargetTransformInfo const&, llvm::DomTreeUpdater*, llvm::SimplifyCFGOptions const&, llvm::ArrayRef<llvm::WeakVH>) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMTransformUtils.so.20.0git+0x2c557b)
#16 0x0000717e635bb663 iterativelySimplifyCFG(llvm::Function&, llvm::TargetTransformInfo const&, llvm::DomTreeUpdater*, llvm::SimplifyCFGOptions const&) SimplifyCFGPass.cpp:0:0
#17 0x0000717e635bc6d4 simplifyFunctionCFGImpl(llvm::Function&, llvm::TargetTransformInfo const&, llvm::DominatorTree*, llvm::SimplifyCFGOptions const&) SimplifyCFGPass.cpp:0:0
#18 0x0000717e635bcf05 simplifyFunctionCFG(llvm::Function&, llvm::TargetTransformInfo const&, llvm::DominatorTree*, llvm::SimplifyCFGOptions const&) SimplifyCFGPass.cpp:0:0
#19 0x0000717e635bd068 llvm::SimplifyCFGPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.20.0git+0x3bd068)
#20 0x0000717e64ed4675 llvm::detail::PassModel<llvm::Function, llvm::SimplifyCFGPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.20.0git+0xd4675)
#21 0x0000717e6215576d llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x35576d)
#22 0x0000717e684d8695 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.20.0git+0xd8695)
#23 0x0000717e621538a6 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x3538a6)
#24 0x0000717e684d9055 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.20.0git+0xd9055)
#25 0x0000717e6215442a llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x35442a)
#26 0x0000717e69b161c1 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x2d1c1)
#27 0x0000717e69b21944 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x38944)
#28 0x0000717e69229d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#29 0x0000717e69229e40 call_init ./csu/../csu/libc-start.c:128:20
#30 0x0000717e69229e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#31 0x00006204288e7095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

**Comment by nikic** [comment]
@dtcxzyw Thanks for pointing that out! I think truncation here is the semantically wrong thing to do. I added a check that we fit the bit width instead in https://github.com/llvm/llvm-project/commit/6f194a6dea4b4067336431e699ea3588417d4b96.

**Comment by nikic** [comment]
Ping. The PR no longer contains MLIR changes now.

**Review by dtcxzyw — COMMENTED** [review]
Middle-end/CodeGen/RISC-V changes LGTM.

File: llvm/lib/Target/Hexagon/HexagonGenExtract.cpp [hunk 5]
```diff
@@ -171,7 +171,7 @@ bool HexagonGenExtract::convert(Instruction *In) {
     // this value.
     if (!LogicalSR && (SR > SL))
       return false;
-    APInt A = APInt(BW, ~0ULL).lshr(SR).shl(SL);
+    APInt A = APInt(BW, ~0ULL, true).lshr(SR).shl(SL);
```
Comment (RKSimon) [inline llvm/lib/Target/Hexagon/HexagonGenExtract.cpp:174]: can we not use APInt::getBitsSet here?

**Comment by dtcxzyw** [comment]
`APInt::getAllOnes` will assert if `BitWidth == 0`.
```
[----------] 6 tests from APFloatTest
[ RUN      ] APFloatTest.MinimumNumber
[       OK ] APFloatTest.MinimumNumber (0 ms)
[ RUN      ] APFloatTest.Float8E8M0FNUValues
ADTTests: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/include/llvm/ADT/APInt.h:116: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `val == 0 && "Value must be zero for 0-bit APInt"' failed.
 #0 0x00007ffff7e15c32 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0x215c32)
 #1 0x00007ffff7e12aff llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0x212aff)
 #2 0x00007ffff7e12c45 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007ffff7442520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007ffff74969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007ffff74969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007ffff74969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007ffff7442476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007ffff74287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007ffff742871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007ffff7439e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007ffff7c97cb6 (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0x97cb6)
#12 0x00007ffff7c98b83 llvm::detail::IEEEFloat::makeNaN(bool, bool, llvm::APInt const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0x98b83)
#13 0x00007ffff7c9e019 llvm::detail::IEEEFloat::convertFromStringSpecials(llvm::StringRef) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0x9e019)
#14 0x00007ffff7caef9d llvm::detail::IEEEFloat::convertFromString(llvm::StringRef, llvm::RoundingMode) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0xaef9d)
#15 0x00007ffff7caf4f6 llvm::APFloat::APFloat(llvm::fltSemantics const&, llvm::StringRef) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libLLVMSupport.so.20.0git+0xaf4f6)
#16 0x000055555585e240 (anonymous namespace)::APFloatTest_Float8E8M0FNUValues_Test::TestBody() APFloatTest.cpp:0:0
#17 0x00007ffff7f6e9d1 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libllvm_gtest.so.20.0git+0x569d1)
#18 0x00007ffff7f6ef2a testing::Test::Run() (.part.0) gtest-all.cc:0:0
#19 0x00007ffff7f6fc02 testing::TestInfo::Run() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libllvm_gtest.so.20.0git+0x57c02)
#20 0x00007ffff7f721e9 testing::TestSuite::Run() (.part.0) gtest-all.cc:0:0
#21 0x00007ffff7f7c18b testing::internal::UnitTestImpl::RunAllTests() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libllvm_gtest.so.20.0git+0x6418b)
#22 0x00007ffff7f6e441 testing::UnitTest::Run() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libllvm_gtest.so.20.0git+0x56441)
#23 0x00007ffff7fb71cc main (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/lib/libllvm_gtest_main.so.20.0git+0x11cc)
#24 0x00007ffff7429d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#25 0x00007ffff7429e40 call_init ./csu/../csu/libc-start.c:128:20
#26 0x00007ffff7429e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#27 0x00005555557b7075 _start (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/unittests/ADT/./ADTTests+0x263075)
```
Would be better to handle this corner case in `IEEEFloat::makeNaN`:
https://github.com/llvm/llvm-project/blob/81fee740d073640c79c0d45a8e8e804c1c5fea40/llvm/lib/Support/APFloat.cpp#L966

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 6]
```diff
@@ -1159,7 +1159,9 @@ class ARMOperand : public MCParsedAsmOperand {
     if (!isImm()) return false;
     const MCConstantExpr *CE = dyn_cast<MCConstantExpr>(getImm());
     if (!CE) return false;
```
Comment (davemgreen) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:1161]: Can this check that CE->getActiveBits() > 32?

**Comment by nikic** [comment]
> `APInt::getAllOnes` will assert if `BitWidth == 0`.
> 
> [snip]

For the record, this is fixed by https://github.com/llvm/llvm-project/pull/112227.

File: llvm/lib/Target/Hexagon/HexagonGenExtract.cpp (hunk 5 above)
Comment (nikic) [reply llvm/lib/Target/Hexagon/HexagonGenExtract.cpp:174]: I tried this, but it's pretty awkward. Note that the shifts are by different values, so this is not just a simple high bits pattern. Doing this as two shifts directly matches the pattern that's being matched above.

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp (hunk 6 above)
Comment (nikic) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp]: Not quite, but this put me on the right path. I added a check for `isUInt<32>` and dropped the signed flag.

File: clang/lib/Parse/ParseInit.cpp [hunk 7]
```diff
@@ -437,7 +437,9 @@ ExprResult Parser::createEmbedExpr() {
   SourceLocation StartLoc = ConsumeAnnotationToken();
   if (Data->BinaryData.size() == 1) {
     Res = IntegerLiteral::Create(Context,
-                                 llvm::APInt(CHAR_BIT, Data->BinaryData.back()),
+                                 llvm::APInt(CHAR_BIT, Data->BinaryData.back(),
```
Comment (efriedma-quic) [inline clang/lib/Parse/ParseInit.cpp:440]: `llvm::APInt(CHAR_BIT, (uint8_t)Data->BinaryData.back())`, maybe?

File: clang/lib/Sema/SemaExpr.cpp [hunk 8]
```diff
@@ -3600,8 +3600,11 @@ ExprResult Sema::ActOnCharacterConstant(const Token &Tok, Scope *UDLScope) {
 
 ExprResult Sema::ActOnIntegerConstant(SourceLocation Loc, uint64_t Val) {
   unsigned IntSize = Context.getTargetInfo().getIntWidth();
-  return IntegerLiteral::Create(Context, llvm::APInt(IntSize, Val),
-                                Context.IntTy, Loc);
+  // TODO: Avoid implicit trunc?
+  return IntegerLiteral::Create(
+      Context,
+      llvm::APInt(IntSize, Val, /*isSigned=*/false, /*implicitTrunc=*/true),
```
Comment (efriedma-quic) [inline clang/lib/Sema/SemaExpr.cpp:3606]: It looks like some of the callers are passing in -1... probably the signature needs to be reworked.  Okay to leave it for now.

File: clang/lib/Parse/ParseInit.cpp (hunk 7 above)
Comment (nikic) [reply clang/lib/Parse/ParseInit.cpp]: Done, though I used `(unsigned char)` to match `CHAR_BIT` more closely.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 9]
```diff
@@ -28729,7 +28731,7 @@ static SDValue GenerateFixedLengthSVETBL(SDValue Op, SDValue Op1, SDValue Op2,
   unsigned BitsPerElt = VTOp1.getVectorElementType().getSizeInBits();
   unsigned IndexLen = MinSVESize / BitsPerElt;
   unsigned ElementsPerVectorReg = VTOp1.getVectorNumElements();
-  uint64_t MaxOffset = APInt(BitsPerElt, -1, false).getZExtValue();
+  uint64_t MaxOffset = APInt(BitsPerElt, -1, true).getZExtValue();
```
Comment (RKSimon) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:28734]: Seems convoluted - `APInt::getAllOnes(BitsPerElt).getZExtValue()` or just do a `~0ULL >> (64 - BitsPerElt)`?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 9 above)
Comment (nikic) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:28734]: After poking around in MathExtras a bit, I replaced this with `maxUIntN(BitsPerElt)`.

**Pull request merged**